### ISSUE TITLE
Terratest fails to run destroy when it cant successfully apply!

### ISSUE THREAD

**Issue by ma-kadri** [gruntwork-io/terratest#511]
I have a scenario where a old test ran and it left behind some stuff in AWS like a instance Profile, Now on the latest run of this test, since while doing a "terraform apply" it failed because of resource exist, Terratest does not runs "terraform destroy" like it does when it is successful at creating "terraform apply" 

Although i have this in the test case.

```
	// At the end of the test, run 'terraform destroy' to clean up any resources that were created.
	defer terraform.Destroy(t, terraformOptions)

```
Error:
```
173 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: Error: Error applying plan:
174 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
175 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 2 errors occurred:
176 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 	* module.ecs-cluster.aws_iam_role.instance: 1 error occurred:
177 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 	* aws_iam_role.instance: Error creating IAM Role rqm-instance-role-test: EntityAlreadyExists: Role with name rqm-instance-role-test already exists.
178 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 	status code: 409, request id: 64540f8d-59d6-449d-b63d-89569c2a39da
179 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
180 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
181 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 	* module.ecs-cluster.aws_iam_role.cadvisor: 1 error occurred:
182 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 	* aws_iam_role.cadvisor: Error creating IAM Role rqm-cadvisor-test: EntityAlreadyExists: Role with name rqm-cadvisor-test already exists.
183 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 	status code: 409, request id: 82804a31-249b-4eac-b28e-a9808523303d
184 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
185 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
186 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
187 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
188 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
189 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: Terraform does not automatically rollback in the face of errors.
190 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: Instead, your Terraform state file has been partially updated with
191 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: any resources that successfully completed. Please address the error
192 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: above and apply again to incrementally change your infrastructure.
193 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
194 TestTerraformAwsEcs 2020-04-21T09:39:36Z command.go:158: 
195 TestTerraformAwsEcs 2020-04-21T09:39:36Z retry.go:80: Returning due to fatal error: FatalError{Underlying: exit status 1}
196 --- FAIL: TestTerraformAwsEcs (93.72s)
197 	apply.go:13: 
198 			Error Trace:	apply.go:13
199 			            				integration_test.go:44
200 			Error:      	Received unexpected error:
201 			            	FatalError{Underlying: exit status 1}
202 			Test:       	TestTerraformAwsEcs
203 FAIL
204 FAIL	test_cases	93.729s
205
Running after_script
00:01
206
Uploading artifacts for failed job
```

**Comment by brikis98** [comment]
Did you call `defer terraform.Destroy(t, terraformOptions)` before or after `terraform.Apply(...)`? It must be called *before* due to how `defer` works in Go.

**Comment by JuananOc** [comment]
Hi, @brikis98 i have the same error, and I have the defer code before the terraform.Apply method.
When I have any problem in the apply section, it will not run any code after

**Comment by brikis98** [comment]
Please share your code so we can see what's happening!

**Comment by arzietlow** [comment]
Hi @brikis98, I'm facing either the same or a similar issue. 

TL;DR : I may be using CopyTerraformFolderToTemp improperly (?)

I have 3 *_test.go files being tested in parallel-- all testing the same module code-- each with a very similar test function which I'll include here:



`package test

import (
	"fmt"
	"strings"
	"testing"
	"github.com/gruntwork-io/terratest/modules/random"
	"github.com/gruntwork-io/terratest/modules/terraform"
	"github.com/gruntwork-io/terratest/modules/test-structure"
)

func TestBurst(t *testing.T) {
	t.Parallel()
    
    rootFolder := ".."
    burstFolderRelativeToRoot := "unit-tests/test-scenarios/burst"
    burstTestFolder := test_structure.CopyTerraformFolderToTemp(t, rootFolder, burstFolderRelativeToRoot)
    
    burstID := random.UniqueId()
	burstEnv := fmt.Sprintf("efs-module-burst-test-%s", strings.ToLower(burstID))
	acctNumber := "XXXXXXXXXXXXXXXX"

	terraformOptionsBurst := &terraform.Options{
	    TerraformBinary: "terraform12",
		TerraformDir: burstTestFolder,
		Vars: map[string]interface{}{
		    "env_name":       burstEnv,
		    "acct_number":    acctNumber,
		    "acct_name":      "XXXXXXXXX",
			"vpc_name":       "XXXXXXXXX",
		},
	}
	
	defer terraform.Destroy(t, terraformOptionsBurst)
	
	terraform.InitAndApply(t, terraformOptionsBurst)

	burst_ID := terraform.Output(t, terraformOptionsBurst, "efs_id")
	burst_ARN := terraform.Output(t, terraformOptionsBurst, "efs_arn")
	
	fmt.Println("BurstID and burstARN:\n" + burst_ID + "\n" + burst_ARN)
	
}
`

I basically have 3 copies of this function in 3 different *_test.go files. Each of them points to terraform code that calls my EFS module (in this case at "unit-tests/test-scenarios/burst"), with each call to that module being slightly different but all using the same **source** path to the actual module code. 

What happens is:
1st run: Applies are all successful. On destroy, one test gives `Error delete file system: fs-xxxxxx with err FileSystemNotFound: File system 'fs-xxxxxx' does not exist.` while the other 2 report successful destroys of the same number of resources that were created in apply. The problem is, the two EFS resources for the 2 "successful" tests are still sitting in my AWS account after the Terratest job has finished. 
My hypothesis: Somehow the 2 "passed" tests are destroying the 3rd "failed" test's EFS resource, and leaving both of their EFS instances behind.

2nd run: As expected, get two **FileSystemAlreadyExists** errors for the two tests that previously reported passing, and the tests fail there.

Is there something I'm overlooking with the placement of the .terraform folder? Maybe I'm using CopyTerraformFolderToTemp incorrectly.. 

Also none of the SKIP_<stage> environment variables are set..

I hope I was able to word this well enough. Greatly appreciate any insight! Thank you!

**Comment by brikis98** [comment]
@arzietlow Your use of `CopyTerraformFolderToTemp ` seems OK at a glance. A few questions:

1. Are you setting `creation_token` in EFS to the same value by any chance? Or any other param that should be unique, but as you're running the same module, may end up getting the same value in all 3 of your concurrent tests?
1. Do you have a `backend` configured for that Terraform code? If so, then even with 3 copies of the module (1 for each test) you're still writing the state file to the same place.

**Comment by arzietlow** [comment]
@brikis98 Thank you for such a fast response! After checking some things, I think #2 must have been my problem. We were originally trying to use a s3 backend along with the unique env_name parameter to make unique locations for each test to use.. but after closer inspection the setup was wrong and all 3 tests were just going to the same place in the bucket every run.. We removed the backend and everything's working nicely. 

Just wanted to thank you again for your help! I greatly appreciate it and the work you do.

**Comment by brikis98** [comment]
Understood, thanks for following up!

**Comment by JuananOc** [comment]
Hi @brikis98 , when reviewing my code to share with you, I saw that efectively I have the defer statement *after* the Apply method, my fault :(

**Comment by brikis98** [comment]
@yorinasub17 Why re-open this one?

**Comment by yorinasub17** [comment]
The last comment was 5 days ago, but it was closed by `no-response` bot. I wanted to make sure we closed it for the right reason.

**Comment by brikis98** [comment]
Hm, OK. Well, given this response, I think it's safe to close this issue:

> Hi @brikis98 , when reviewing my code to share with you, I saw that efectively I have the defer statement _after_ the Apply method, my fault :(
